Optimization and Cleanup

terminalforlife opened the pull request:
Hi. See commit message for information on the changes made.

Commit:
Optimize and de-clutter

I'd ideally like to sort (reverse) with `awk`, and it's doable, but
unfortunately I can't remember how to do it, and my search online isn't
yielding any results.

It's a real shame to have to re-execute `awk` each line, so I will
probably revisit this block, because I feel like it could certainly be
further optimized. This is still, however, absolutely an improvement.

If there are any issues with this block, please let me know, so that I
might hopefully address them.

inferont commented:
Nice! I'll test this out this week.

inferont reviewed (changes requested):
This definitely is much faster! It looks like sorting the branches by date descending is broken though. Would you be able to fix that?

ghost commented:
Gooood. I'll give it a go, now.

Commit:
Address request in #1 and make code leaner

If I have the sorting the wrong way around, just remove the `-r` flag
by the `sort` command. I've omitted AWK, releasing a process and saving
some resources. Any speed loss for this change _should_ be negligible.

Slightly off-topic, but because people live all over the world, I
recommend that you use the time zone thing (`+0100`, for example), -
otherwise it'll be misleading as to _when_ somebody _really_ committed
those changes. Food for thought.

Is there anything else at all that you'd like done with this?

terminalforlife commented on branches:
```diff
@@ -1,11 +1,7 @@
+while read Branch _; do
+	read EMail Date Time _ <<< "$($GitCMD log --format='%ae %ci' $Branch)"
+	printf "%s %s %s %s\n" "${Branch##*/}" "$EMail" "$Date" "$Time"
+done <<< "$($GitCMD branch -r --no-color)" | LC_ALL='C' sort -rk 3,4 | column -tx
```
By the way, I used `LC_ALL='C'` in order to avoid locale-specific sorting, so as to _helpfully_ get a more consisting and expecting sorting experience.

Commit:
Add more content to `README.md`

Changes:

  * Minor English improvements.

  * Improved installation instructions.

  * Added command for backing up the `.bashrc` file.

  * Added a section for restoring the backup (`.bashrc.old`).

  * Added Curl as an alternative, if the user doesn't have Wget.

Commit:
Correct typo

inferont commented:
The documentation changes are great and I could merge them separately if you'd like. 

I profiled the changes to `branches` on a few different repositories and it looks to be much slower when the branch count is in the thousands. I noticed that it was faster on smaller repositories though. I used the `time branches` command with the three different versions of the branches script. With this huge repository the original version scored around 6.6s, the second version you introduced was around 7.2s, and this latest version went into the minutes before I killed the command.

Thanks!

ghost commented:
Wow. Holy crap, it never occurred to me a repository would have _thousands_ of branches. lol No wonder it was slow. I'll revise my previous AWK code if I get the chance. Thanks for getting back to me, and I apologise for the inconvenience.

Oh, and yes if you'd like and are able to merge just the README.md changes for now, you're welcome to.

Question: is Perl an option? I'm writing up a quick Perl script for this, which I expect will be very fast.

Also, I could do with knowing which repository you tested this on, with all those branches. Then I could test it myself and come back to you with a solution which I've observed is faster. Saves you hassle. lol

ghost commented:
Okay, so this is a little project I've started based on what you're wanting, so I'm hoping it helps you as well. As I've written it from scratch, I've got it in one of my own repositories, but since the inspiration came from you and this repository, I've added as much in the header (comments) and a link to this repository.

It's not finished yet, but it's getting there:

https://github.com/terminalforlife/PerlProjects/blob/dev/gitbrov

When I'm happy enough with it (stable, code's acceptable, presentable, etc), I'll start looking into ways I can make it even faster. It'd be interesting to know how it performs so far, though.

inferont commented:
> Question: is Perl an option? I'm writing up a quick Perl script for this, which I expect will be very fast.

I think Perl is workable for the branches command itself or anything related to development work as long as the script uses a default set of Perl libraries. Replacing the `cd` command with something that uses Perl would be problematic though. Besides making the daily life of myself or others in a shell easier, the goal of these scripts is to be unified when connecting to dozens of different servers a day.

> Also, I could do with knowing which repository you tested this on, with all those branches.

I can not really share this repository since it's private. It also seems like most open source projects do some housekeeping with their branches as I can't find a public example here. :|

I can manually test anything you've written until we can find some repo with a lot of branches or maybe decide to fake one.

ghost commented:
Okidoke. So my lil Perl project might not be ideal, as it depends on `libgit-perl` to communicate directly with git, which doesn't need cd-ing all over the place. I could have Perl run git via shell commands, but it's sloppy and I have to wonder if there are security implications of it.

I could try parsing files in `.git` since it'd be fast as heck, but it'll take some time for me to understand what really goes on in `.git`; how it's organised and how they might be parsed. I'm actually in the process of reading a book on git which covers a whole crap-ton of stuff, so hopefully it goes into detail there.

By the way, the git commands take `-C [PATH]` as an argument. Found that out a few days ago. It's very useful! Saves the cd-ing.

By the way, Perl can cd as well, or did you mean something else? Regarding the problematic thing.

Thank you for letting me work with you on this. It's been very educational and a fun challenge.

inferont commented:
> I could try parsing files in .git since it'd be fast as heck, but it'll take some time for me to understand what really goes on in .git; how it's organised and how they might be parsed. 

This sounds like it could be incredibly fragile. Anything the git project decides to do with these files would require an update and lots of switch statements to handle the versions of git across different servers.

> By the way, Perl can cd as well, or did you mean something else? Regarding the problematic thing.

I was trying to give a feel for what this "shell-scripts" project is about regarding the range of environments it should support. I currently use it on small IoT embedded systems, productions servers where I shouldn't be compiling or installing any package, and a variety of development machines. This branches command would only be used on development machines so I think using Perl here is appropriate. There are some minimal machines that do not have access to Perl so replacing the `cd` command with something that requires Perl would not be workable. (I've replaced the default `cd` behavior when no parameters are specified in .commonrc with a list of common paths.)

ghost commented:
Okay, so Perl is an option for development machines but not the embedded ones? Best I head back to the drawling board. lol I'll get back to you when/if I come up with something (hopefully) useful. :)

inferont commented:
Yep! That's what I was attempting to say. :D

inferont commented:
> Oh, and yes if you'd like and are able to merge just the README.md changes for now, you're welcome to.

Merged the README updates and made some additions. Thanks!

inferont commented:
I will probably attempt to integrate some of the changes you made in your fork this week. A curl/wget agnostic downloader is pretty useful as well as some of the other stuff.

ghost commented:
Awesome. :) Wish I could get involved, but I'm gonna be really busy over the next couple of weeks. If you happen to message though, I will hopefully be able to get to it within a day or two.
